### PR TITLE
Token Get Nft Infos Query implementation

### PR DESCRIPTION
**Pull request by daniel-danev-limechain**
**Related issue(s)**:
Closes #<issue number>

**Summary of the change**:
Lorem ipsum dolor sit amet.

**External impacts**:
None.

**Applicable documentation**
- [ ] Javadoc
- [ ] HAPI protobuf comments
- [ ] README

### CODE REVIEW

**Review by Daniel-K-Ivanov — COMMENTED** [review]
It seems that the License is not the same as the rest of the files in terms of formatting and the text for the name of the project is not correct. The license file in `hapi-fees` is one, in `hedera-node` another and `test-clients` third. Here the license is always the one for `hedera-node` -> notice the start of the license "Hedera Services Node"

File: hedera-node/src/main/java/com/hedera/services/queries/token/GetTokenNftInfosAnswer.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+	}
+
+	@Override
+	public Response responseGiven(
+			Query query, StateView view, ResponseCodeEnum validity, long cost, Map<String, Object> queryCtx) {
+		return responseFor(query, view, validity, cost, Optional.of(queryCtx));
+	}
+
+	@Override
+	public ResponseCodeEnum checkValidity(Query query, StateView view) {
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/queries/token/GetTokenNftInfosAnswer.java:79]: `QUERY_RANGE_LIMIT_EXCEEDED` is misleading response when `start` > `end` or `start < 0 || end < 0`. @georgiyazovaliiski renamed the response code to a more appropriate one. Use that instead

File: hedera-node/src/main/java/com/hedera/services/state/merkle/MerkleUniqueToken.java [hunk 2]
```diff
@@ -62,6 +62,10 @@ public MerkleUniqueToken() {
 		/* No-op. */
 	}
 
+	public MerkleUniqueToken(EntityId owner) {
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/state/merkle/MerkleUniqueToken.java:65]: Is this change necessary?

File: hedera-node/src/main/java/com/hedera/services/store/tokens/ExceptionalTokenStore.java [hunk 3]
```diff
@@ -26,6 +26,7 @@
 import com.hedera.services.state.merkle.MerkleAccount;
 import com.hedera.services.state.merkle.MerkleToken;
 import com.hedera.services.state.merkle.MerkleUniqueToken;
+import com.hedera.services.state.merkle.MerkleUniqueTokenId;
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/ExceptionalTokenStore.java:29]: Is this change necessary?

File: hedera-node/src/main/java/com/hedera/services/store/tokens/ExceptionalTokenStore.java [hunk 4]
```diff
@@ -34,6 +35,7 @@
 import com.hederahashgraph.api.proto.java.TokenCreateTransactionBody;
 import com.hederahashgraph.api.proto.java.TokenID;
 import com.hederahashgraph.api.proto.java.TokenUpdateTransactionBody;
+import org.apache.commons.lang3.tuple.Pair;
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/ExceptionalTokenStore.java:38]: Is this change necessary?

File: hedera-node/src/main/java/com/hedera/services/store/tokens/TokenStore.java [hunk 5]
```diff
@@ -23,6 +23,7 @@
 import com.hedera.services.ledger.BalanceChange;
 import com.hedera.services.state.merkle.MerkleToken;
 import com.hedera.services.state.merkle.MerkleUniqueToken;
+import com.hedera.services.state.merkle.MerkleUniqueTokenId;
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/TokenStore.java:26]: Is this change necessary?

File: hedera-node/src/main/java/com/hedera/services/store/tokens/TokenStore.java [hunk 6]
```diff
@@ -32,6 +33,7 @@
 import com.hederahashgraph.api.proto.java.TokenCreateTransactionBody;
 import com.hederahashgraph.api.proto.java.TokenID;
 import com.hederahashgraph.api.proto.java.TokenUpdateTransactionBody;
+import org.apache.commons.lang3.tuple.Pair;
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/store/tokens/TokenStore.java:36]: Is this change necessary?

File: hedera-node/src/main/java/com/hedera/services/context/primitives/StateView.java [hunk 7]
```diff
@@ -403,6 +404,28 @@ public StateView(
 		return Optional.of(info);
 	}
 
+	public Optional<List<TokenNftInfo>> infoForTokenNfts(TokenID tid, long start, long end) {
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/context/primitives/StateView.java:407]: There is no unit test coverage for the method

File: hedera-node/src/main/java/com/hedera/services/queries/token/TokenAnswers.java [hunk 8]
```diff
@@ -23,15 +23,23 @@
...
 	}
 
-	public GetTokenNftInfoAnswer getNftInfoAnswer() { return nftInfo; }
+	public GetTokenNftInfoAnswer getNftInfoAnswer() {
+		return nftInfo;
+	}
+
+	public GetTokenNftInfosAnswer getTokenNftInfosAnswer() {
+		return tokenNftInfos;
+	}
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/queries/token/TokenAnswers.java:44]: No unit test coverage

File: hedera-node/src/test/java/com/hedera/services/queries/token/GetTokenNftInfosAnswerTest.java [hunk 9]
```diff
@@ -0,0 +1,337 @@
...
+	GetTokenNftInfosAnswer subject;
+	Transaction paymentTxn;
+	StateView view;
+	private long fee = 1_234L;
+
+	String node = "0.0.3";
+	String payer = "0.0.1";
+	TokenID tokenId = asToken("0.0.2");
+	TokenID invalidTokenId = asToken("0.0.4");
+	FCMap<MerkleEntityId, MerkleAccount> accountMap;
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/test/java/com/hedera/services/queries/token/GetTokenNftInfosAnswerTest.java:75]: Is this necessary?

File: hedera-node/src/test/java/com/hedera/services/queries/token/GetTokenNftInfosAnswerTest.java [hunk 10]
```diff
@@ -0,0 +1,337 @@
...
+				TokenNftInfo.newBuilder()
+						.setMetadata(ByteString.copyFromUtf8("stuff3"))
+						.build()
+		));
+		accountMap = new FCMap<>();
+		accountMap.put(new MerkleEntityId(
+				tokenId.getShardNum(),
+				tokenId.getRealmNum(),
+				tokenId.getTokenNum()
+		), new MerkleAccount(new ArrayList<>()));
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/test/java/com/hedera/services/queries/token/GetTokenNftInfosAnswerTest.java:104]: Is this necessary?

File: hedera-node/src/test/java/com/hedera/services/txns/validation/ContextOptionValidatorTest.java [hunk 11]
```diff
@@ -604,8 +607,17 @@ void memoCheckWorks() {
 	}
 
 	@Test
-	void rejectsInvalidBatchSize(){
+	void rejectsInvalidBatchSize() {
 		var list = LongStream.range(0, 1000).boxed().collect(Collectors.toList());
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/test/java/com/hedera/services/txns/validation/ContextOptionValidatorTest.java:611]: Do we need the `list`variable?

File: hedera-node/src/test/java/com/hedera/services/txns/validation/ContextOptionValidatorTest.java [hunk 12]
```diff
@@ -604,8 +607,17 @@ void memoCheckWorks() {
...
+		assertEquals(BATCH_SIZE_LIMIT_EXCEEDED, subject.maxBatchSizeMintCheck(list.size()));
+		assertEquals(BATCH_SIZE_LIMIT_EXCEEDED, subject.maxBatchSizeWipeCheck(list.size()));
+		assertEquals(BATCH_SIZE_LIMIT_EXCEEDED, subject.maxNftTransfersLenCheck(list.size()));
+	}
+
+	@Test
+	void rejectsInvalidMetadata() {
+		assertEquals(QUERY_RANGE_LIMIT_EXCEEDED, subject.nftMaxQueryRangeCheck(0, 1001));
+		assertEquals(METADATA_TOO_LONG, subject.nftMetadataCheck(ByteString.copyFromUtf8("aaaaaaaaaaaaaaa").toByteArray()));
 	}
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/test/java/com/hedera/services/txns/validation/ContextOptionValidatorTest.java:622]: You must mock the `dynamicProperties`. Atm they are `0`. We have to use some sensible number

File: test-clients/src/main/java/com/hedera/services/bdd/spec/queries/token/HapiGetTokenNftInfos.java [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+
+	public HapiGetTokenNftInfos(String token, long start, long end) {
+		this.token = token;
+		this.start = start;
+		this.end = end;
+	}
+
+	@Override
+	public HederaFunctionality type() {
+		return HederaFunctionality.TokenGetNftInfo;
```
Comment (Daniel-K-Ivanov) [inline test-clients/src/main/java/com/hedera/services/bdd/spec/queries/token/HapiGetTokenNftInfos.java:65]: This is not correct

File: test-clients/src/main/java/com/hedera/services/bdd/suites/token/UniqueTokenManagementSpecs.java [hunk 14]
```diff
@@ -550,6 +559,129 @@ private HapiApiSpec uniqueTokenMintReceiptCheck() {
...
+	public HapiApiSpec failsWithFungibleTokenGetNftInfos() {
+		return defaultHapiSpec("FailsWithFungibleTokenGetNftInfos")
+				.given(
+						cryptoCreate(TOKEN_TREASURY).balance(0L)
+				).when(
+						tokenCreate(A_TOKEN)
+								.initialSupply(1_000)
+								.treasury(TOKEN_TREASURY)
+				).then(
+						getTokenNftInfos(A_TOKEN, 0, 2)
```
Comment (Daniel-K-Ivanov) [inline test-clients/src/main/java/com/hedera/services/bdd/suites/token/UniqueTokenManagementSpecs.java:680]: The response code for this query is not appropriate

**Review by Daniel-K-Ivanov — CHANGES_REQUESTED** [review]
You can add the new query to the rest of the transfer/mint/burn operations to verify that the FCOTMR association is updated correctly on transfer/mint/burn

**Review by joan41868 — APPROVED** [review]

File: test-clients/src/main/java/com/hedera/services/bdd/suites/token/UniqueTokenManagementSpecs.java (hunk 14 above)
Comment (daniel-danev-limechain) [reply test-clients/src/main/java/com/hedera/services/bdd/suites/token/UniqueTokenManagementSpecs.java:681]: I've traversed all the available responses and to my mind NOT_SUPPORTED is the best choice here

File: hedera-node/src/main/java/com/hedera/services/queries/token/GetTokenNftInfosAnswer.java [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+		var tokenNftInfosQuery = query.getTokenGetNftInfos();
+		var id = tokenNftInfosQuery.getTokenID();
+
+		var optionalInfo = view.infoForToken(id);
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/queries/token/GetTokenNftInfosAnswer.java:85]: `infoForToken` is pretty big operation, populating a lot of stuff. I think that it would be better to have a lightweight check. There is `tokenExists` method, but unfortunately, it only checks for existence. Ideally we would like to check for the existence but include the check for deleted as-well.

I think that it would be best to use the `tokenWith` method and operate on that

File: hedera-node/src/main/java/com/hedera/services/queries/token/GetTokenNftInfosAnswer.java [hunk 16]
```diff
@@ -0,0 +1,187 @@
...
+
+		if (tokenNftInfosQuery.getStart() >= tokenNftInfosQuery.getEnd()) {
+			return INVALID_QUERY_RANGE;
+		}
+
+		var validity = validator.nftMaxQueryRangeCheck(
+				tokenNftInfosQuery.getStart(), tokenNftInfosQuery.getEnd());
+		if (validity != OK) {
+			return validity;
+		}
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/queries/token/GetTokenNftInfosAnswer.java:107]: Failing fast if there is an error is a priority, that being said I think that it is better to place those 2 checks before we load the Token from the state

File: hedera-node/src/main/java/com/hedera/services/queries/token/GetTokenNftInfosAnswer.java [hunk 17]
```diff
@@ -0,0 +1,187 @@
...
+				setAnswerOnly(response, view, op, cost, queryCtx);
+			}
+		}
+
+		return Response.newBuilder()
+				.setTokenGetNftInfos(response)
+				.build();
+	}
+
+	@SuppressWarnings("unchecked")
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/queries/token/GetTokenNftInfosAnswer.java:160]: remove @suppress warnings

File: hedera-node/src/main/java/com/hedera/services/state/submerkle/EntityId.java [hunk 18]
```diff
@@ -190,6 +190,22 @@ public static EntityId fromGrpcContractId(ContractID id) {
...
+				.build();
+	}
+
+	public static TokenID fromEntityTokenId(EntityId e) {
+		return TokenID.newBuilder()
+				.setRealmNum(e.realm())
+				.setShardNum(e.shard())
+				.setTokenNum(e.num())
+				.build();
+	}
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/state/submerkle/EntityId.java:207]: Can't we use `toGrpcAccountId` and `toGrpcTokenId`

File: hedera-node/src/test/java/com/hedera/services/context/ServicesContextTest.java [hunk 19]
```diff
@@ -122,6 +122,7 @@
 import com.hedera.services.state.migration.StdStateMigrations;
 import com.hedera.services.state.submerkle.EntityId;
 import com.hedera.services.state.submerkle.ExchangeRates;
+import com.hedera.services.state.submerkle.RichInstant;
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/test/java/com/hedera/services/context/ServicesContextTest.java:125]: unnecessary import

File: test-clients/src/main/java/com/hedera/services/bdd/spec/queries/token/HapiGetTokenNftInfos.java [hunk 20]
```diff
@@ -0,0 +1,134 @@
...
+					expectedNftElement.setCreationTime(spec.registry().getCreationTime(e));
+				});
+				nftInfo.getExpectedSerialNum().ifPresent(expectedNftId::setSerialNumber);
+
+				expectedNftElement.setNftID(expectedNftId.build());
+				nftInfo.getExpectedAccountID().ifPresent(e -> expectedNftElement.setAccountID(TxnUtils.asId(e, spec)));
+				nftInfo.getExpectedMetadata().ifPresent(expectedNftElement::setMetadata);
+
+				var completedNft = expectedNftElement.build();
+				Assert.assertTrue(actualInfo.contains(completedNft));
```
Comment (Daniel-K-Ivanov) [inline test-clients/src/main/java/com/hedera/services/bdd/spec/queries/token/HapiGetTokenNftInfos.java:86]: There is a bug in this implementation. See https://github.com/hashgraph/hedera-services/pull/1724/files#diff-51458346123a0079d1ac75c581029bf8ea203fffcb45dc77eb24f987b964933aR91 for reference. 

https://github.com/hashgraph/hedera-services/pull/1615#discussion_r662168616

**Review by Daniel-K-Ivanov — CHANGES_REQUESTED** [review]
Overall looks good!

File: hedera-node/src/main/java/com/hedera/services/queries/token/GetTokenNftInfosAnswer.java [hunk 21]
```diff
@@ -0,0 +1,186 @@
...
+			return INVALID_QUERY_RANGE;
+		}
+
+		var validity = validator.nftMaxQueryRangeCheck(
+				tokenNftInfosQuery.getStart(), tokenNftInfosQuery.getEnd());
+		if (validity != OK) {
+			return validity;
+		}
+
+		var optionalId = view.tokenWith(id);
```
Comment (Daniel-K-Ivanov) [inline hedera-node/src/main/java/com/hedera/services/queries/token/GetTokenNftInfosAnswer.java:95]: I thin that a better name for the variable would be token or optionalToken

**Review by Daniel-K-Ivanov — APPROVED** [review]
